src/firmware.cpp

```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
-    pinMode(pin, OUTPUT);
-    digitalWrite(pin, LOW);
-  } else if (pinStateArg == "hi-z") {
-    pinMode(pin, INPUT);
-  } else if (pinStateArg == "pullup") {
-    pinMode(pin, INPUT_PULLUP);
+static CommandResponse led(int requestID, String argument) {
+  if (argument == "H") {
+    digitalWrite(LED_BUILTIN, HIGH);
+  } else if (argument == "L") {
+    digitalWrite(LED_BUILTIN, LOW);
   } else {
-    return COMMAND_ERROR("unknown drive state");
+    return COMMAND_ERROR("Unknown LED State: " + argument);
   }
   return OK;
 }
 
-static CommandError read_pin(int commandId, String argument) {
+static CommandResponse readPin(int requestID, String argument) {
   String pinIDArg = pop_option(argument);
 
   if (argument.length() || !pinIDArg.length()) {
-    return COMMAND_ERROR("need exactly one argument: <pin>");
+    return COMMAND_ERROR("Bad number of arguments");
   }
 
   int pin = pinIDArg.toInt();
 
-  if (pin < 2 || pin > 12) {
-    return COMMAND_ERROR("pin must be between 2 and 12");
+  if (pin < 2 || pin > 13) {
+    return COMMAND_ERROR("pin must be between 2 and 13");
   }
 
-  auto state = digitalRead(pin);
+  int state = digitalRead(pin);
 
   if (state == HIGH) {
-    serialWrite(commandId, '>', "high");
+    serialWrite(requestID, '>', "H");
   } else {
-    serialWrite(commandId, '>', "low");
+    serialWrite(requestID, '>', "L");
   }
 
   return OK;
 }
 
-static void read_analogue_pin_to_serial(int commandId, const String& name, int pin) {
-  int reading = analogRead(pin);
-  double mungedReading = (double)reading * (5.0 / 1024.0);
-  serialWrite(commandId, '>', name + " " + String(mungedReading));
-}
+static CommandResponse servo(int requestID, String argument) {
+  String servoArg = pop_option(argument);
+  String widthArg = pop_option(argument);
+
+  if (argument.length() || !servoArg.length() || !widthArg.length()) {
+    return COMMAND_ERROR("Bad number of arguments.");
+  }
 
-static CommandError analogue_read(int commandId, String argument) {
-  read_analogue_pin_to_serial(commandId, "a0", A0);
-  read_analogue_pin_to_serial(commandId, "a1", A1);
-  read_analogue_pin_to_serial(commandId, "a2", A2);
-  read_analogue_pin_to_serial(commandId, "a3", A3);
+  auto width = widthArg.toInt();
+  auto servo = servoArg.toInt();
+  if (servo < 0 || servo > 15) {
+    return COMMAND_ERROR("Servo out of range");
+  }
+  if (width != 0 && (width < 150 || width > 550)) {
+    return COMMAND_ERROR("Width must be 0 or between 150 and 550");
+  }
+  SERVOS.setPWM(servo, 0, width);
   return OK;
 }
 
-static CommandError ultrasound_read(int commandId, String argument) {
+static CommandResponse ultrasoundRead(int requestID, String argument) {
   String triggerPinStr = pop_option(argument);
   String echoPinStr = pop_option(argument);
 
   if (argument.length() || !triggerPinStr.length() || !echoPinStr.length()) {
-    return COMMAND_ERROR("need exactly two arguments: <trigger-pin> <echo-pin>");
+    return COMMAND_ERROR("Bad Arguments");
```

Comment:
I think the old error message worked better

Reply:
I agree that numerical error codes are a better idea. I will switch them back to the previous ones and change to numerical ones if I have time.

Reply:
See 0cfaddd